client/src/components/MediaScanner/MediaScanner.tsx

```diff
@@ -0,0 +1,71 @@
+import React, { useEffect, useState } from "react";
+import { isWeb, microapps } from "src/config";
+import { MediaResponse, emptyMediaResponse } from "src/interfaces";
+import { processImageBarcode } from "src/utils";
+
+function MediaScanner({
+  button,
+  resultCallback,
+  debugFallbackImg,
+}: {
+  button: React.ReactElement;
+  resultCallback: (barcode: string) => void;
+  debugFallbackImg?: string;
+}) {
+  const [uploadedImg, setUploadedImg] = useState<MediaResponse>();
+
+  const uploadedImgId = "media-img-upload";
+  const debugImgId = "media-img-debug";
+
+  const requestMediaUpload = async () => {
+    if (isWeb) {
+      scanBarcodeFromImage();
+    } else {
+      const imgReq = {
+        allowedMimeTypes: ["image/jpeg"],
+        allowedSources: ["camera"],
+      };
+      const imgRes = await microapps.requestMedia(imgReq);
+      setUploadedImg(imgRes);
+    }
+  };
+
+  const scanBarcodeFromImage = () => {
+    let img = document.getElementById(uploadedImgId) as HTMLImageElement;
+    // Try to fallback, else abort
+    if (!img && debugFallbackImg) {
+      img = document.getElementById(debugImgId) as HTMLImageElement;
+    } else {
+      return;
+    }
+    processImageBarcode(img).then((barcode: string) => {
+      resultCallback(barcode);
+    });
+  };
+
+  useEffect(() => {
+    if (uploadedImg) {
+      scanBarcodeFromImage();
+    }
+  }, [uploadedImg]);
+
+  return (
+    <div className="MediaScanner">
+      {React.cloneElement(button, {
+        onClick: requestMediaUpload,
+      })}
```

Comment:
why not use a child element with onClick to a user-defined value?
where else is the media scanner re-used?

Reply:
My intention was to add a 'standard' onClick (internal-to-MediaScanner `requestMediaUpload` function) to whatever element the user passes to this component as the 'clickable' object to trigger uploading media flow

Reply:
Then you should use this approach:

`
<Media Scanne props...> 
// Whatever child element you want to add here
     <Button onClick="func()"/>
</MediaScanner>
`

The child elements should be accessible via props.children or something similar.

Reply:
Pattern is to leave styling to the parent component/page this component is used in and simply co-opt the button as a trigger for the rest of our media flow. Ideally, I would just add an `onClick` to the props for passed `button` ReactElement but cloning is necessary since props are frozen.

Reply:
yup i mean, for this use case you should handle like i mentioned here:
```
<Media Scanne props...>    
// Whatever child element you want to add here
     <Button onClick="func()"/>
</MediaScanner>
```

This allows you to add any child element easily with whatever external props or styling you want to associate.

Reply:
I see, I have adopted the following:
```
<MediaScanner>
  <div onClick="func()"> // 'clickable group'
    {button} // render user-specified button here
  </div>
  <div>
    // Other HTML elements to store img info
  </div>
</MediaScanner>
```

Reply:
ah i see! This and your prev solution also works. What i mean actually was not changing it here, but changing it in the place where <MediaScanner> is used.
so e.g.  A simplified example:

```
function MediaScanne({children: React.Component}){
return <div>
<p>This is media scanner with its children you added:</p>
{children}
</div>
}
```

A parent function:
```
function ParentOfScannerWithFancyButton(){
return <MediaScanner props=...>
<Button onClick={take to cart page} />
</MediaScanner>
}
```

Another parent may want a different button:

```
function ParentOfScannerWithFancyButton(){
return <MediaScanner props=...>
<FlatIconButton onClick={take to cart page} />
</MediaScanner>
}
```

It can be something other than button too, but you can type check to make sure only buttons are passed:
```
function ParentOfScannerWithFancyButton(){
return <MediaScanner props=...>
<AnotherCustomComponent onClick={take to cart page} />
</MediaScanner>
}
```

Reply:
https://codeburst.io/a-quick-intro-to-reacts-props-children-cb3d2fce4891